src/Pact/Types/Exp.hs

```diff
@@ -155,50 +149,96 @@ litToPrim LDecimal {} = TyDecimal
 litToPrim LBool {} = TyBool
 litToPrim LTime {} = TyTime
 
--- | Pact expressions, with parsing info.
-data Exp =
-  -- | Literals
-  ELiteral { _eLiteral :: !Literal, _eParsed :: !Parsed } |
-  -- | Symbol, effectively a string literal
-  ESymbol { _eSymbol :: !Text, _eParsed :: !Parsed } |
-  -- | Atom, with support for type literals.
-  EAtom { _eAtom :: !Text
-        , _eQualifier :: !(Maybe Text)
-        , _eType :: !(Maybe (Type TypeName))
-        , _eParsed :: !Parsed
-        } |
-  -- | Lists. '_eLitListType' distinguishes literal lists (`[1 2 3]`) from body forms.
-  EList { _eList :: ![Exp], _eLitListType :: !(Maybe (Type TypeName)), _eParsed :: !Parsed } |
-  -- | Object literals.
-  EObject { _eObject :: ![(Exp,Exp)], _eParsed :: !Parsed } |
-  -- | Special binding forms.
-  EBinding { _eBinding :: ![(Exp,Exp)], _eParsed :: !Parsed }
-           deriving (Eq,Generic)
-
-instance NFData Exp
-
-
-pattern EList' :: [Exp] -> Exp
-pattern EList' ls <- EList ls Nothing _
-pattern ELitList :: [Exp] -> Exp
-pattern ELitList ls <- EList ls (Just _) _
-pattern EAtom' :: Text -> Exp
-pattern EAtom' tag <- EAtom tag Nothing Nothing _
-pattern ELitString :: Text -> Exp
-pattern ELitString s <- ELiteral (LString s) _
-
-makePrisms ''Exp
-
+data ListDelimiter = Parens|Brackets|Braces deriving (Eq,Show,Ord,Generic,Bounded,Enum)
+instance NFData ListDelimiter
+
+listDelims :: ListDelimiter -> (Text,Text)
+listDelims Parens = ("(",")")
+listDelims Brackets = ("[","]")
+listDelims Braces = ("{","}")
+
+enlist :: ListDelimiter -> ((Text,Text) -> a) -> a
+enlist d f = f (listDelims d)
+
+data Separator = Colon|ColonEquals|Comma deriving (Eq,Ord,Generic,Bounded,Enum)
+instance NFData Separator
+instance Show Separator where
+  show Colon = ":"
+  show ColonEquals = ":="
+  show Comma = ","
+
+class ExpInfo b a | a -> b where
```

Comment:
Might be worth monomorphizing `b` to just be `Info`. `ExpInfo` just seems very broad and lawless as-is.

Reply:
The main reason is this parser spits out `Parsed` while ExpParser/Compile needs `Info`.

Reply:
How about a type family?

Reply:
Is there any chance then that this parser could also produce a `Code` alongside its `Parsed`, and then we could simplify the design by monomorphizing to `Info`?

Reply:
You can easily monomorphize. This compiles. https://github.com/kadena-io/pact/commit/1071f54e609535c0b07c5d6e471295375eb6f4fa

Reply:
Cool

Reply:
To clarify, the approach I was hinting at would get rid of the typeclass, but would be a more significant change. It seems like we should probably at least do 1071f54e609535c0b07c5d6e471295375eb6f4fa

Reply:
If we could change it so that both this parser and ExpParser/Compile use the same representation, would that also eliminate the new type parameter from `Exp`?